### PR TITLE
Migrate ESPColor to Color

### PR DESCRIPTION
**Pull request by SenexCrenshaw**
## Description:
Remove ESPColor in favor of Color in color.h. This compiles with neopixel and works with the ST7735 driver in the PR. I moved the high level color conversion functions to a separate header.

This is definitely a POC and if this direction is approved I will clean things up and make sure everything is converted and tests work.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1036
## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/color/__init__.py [hunk 1]
```diff
@@ -2,22 +2,49 @@
...
+    cg.variable(
+        config[CONF_ID],
+        cg.StructInitializer(
+            ColorStruct,
+            ("r", config[CONF_RED]),
+            ("g", config[CONF_GREEN]),
+            ("b", config[CONF_BLUE]),
+            ("w", config[CONF_WHITE]),
+        ),
+    )
```
Comment (glmnet) [inline esphome/components/color/__init__.py:50]: please avoid format only changes, if this is automated review your tools

File: 1 [hunk 2]
```diff
@@ -0,0 +1,2039 @@
+substitutions:
```
Comment (glmnet) [inline 1:1]: whats this??

File: esphome/components/color/__init__.py [hunk 3]
```diff
@@ -2,22 +2,49 @@
...
+        raise cv.Invalid(msg)
+    if value < 0:
+        raise cv.Invalid("Value must not be less than 0.")
+    return value
+
+
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.declare_id(ColorStruct),
+        cv.Optional(CONF_RED, default=0): percentage_or_uint8_t,
```
Comment (glmnet) [inline esphome/components/color/__init__.py:32]: this is not good to accept percentage or ints with different values.

I believe percentage already accepts decimals, e.g. `0.43` so now that will mean 0 when converted to int in line 17

You cann add another constructor though, like ARGB=[0x]AABBCCxx

see cv.Exclusive

File: esphome/core/color_utils.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "color.h"
+
+namespace esphome {
```
Comment (glmnet) [inline esphome/core/color_utils.h:5]: should these converters belong to display namespace?

also util is too generic IMO

File: tests/test1.yaml [hunk 5]
```diff
@@ -1683,7 +1683,7 @@ interval:
 color:
   - id: kbx_red
     red: 100%
-    green: 1%
+    green: 123
```
Comment (glmnet) [inline tests/test1.yaml:1686]: this is weird, I don't think we want to accept this

File: esphome/components/light/effects.py [hunk 6]
```diff
@@ -1,38 +1,69 @@
...
+    CONF_BLUE,
+    CONF_WHITE,
+    CONF_ALPHA,
+    CONF_INTENSITY,
+    CONF_SPEED,
+    CONF_WIDTH,
+    CONF_NUM_LEDS,
+    CONF_RANDOM,
+    CONF_SEQUENCE,
+)
```
Comment (glmnet) [inline esphome/components/light/effects.py:24]: even though this formatter is better imo, reviewing the pr with changes like this is almost impossible.
A format change should be in a saparate PR (I am NOT asking for that)

I'll expect all format changes reverted. 

Thanks

File: esphome/components/color/__init__.py (hunk 1 above)
Comment (SenexCrenshaw) [reply esphome/components/color/__init__.py:50]: reformatted

File: 1 (hunk 2 above)
Comment (SenexCrenshaw) [reply 1:1]: Test file, removed

File: esphome/components/color/__init__.py (hunk 3 above)
Comment (SenexCrenshaw) [reply esphome/components/color/__init__.py:32]: I rewrote this

File: tests/test1.yaml (hunk 5 above)
Comment (SenexCrenshaw) [reply tests/test1.yaml:1686]: I added _int counterparts

File: esphome/components/light/effects.py (hunk 6 above)
Comment (SenexCrenshaw) [reply esphome/components/light/effects.py:24]: Yes i was using black, everything is reverted

File: esphome/core/color_utils.h (hunk 4 above)
Comment (SenexCrenshaw) [reply esphome/core/color_utils.h:5]: Yes, this has been moved

**Review by glmnet — APPROVED** [review]
@kbx81 is this ok with you?

**Comment by kbx81** [comment]
Wow...I didn't realize there was quite so much going on here -- but yes, absolutely @glmnet! 😄 I think the consolidation makes sense; I may have to change some of my code but I am happy to do so. @SenexCrenshaw and I have also discussed it a fair bit on discord. 😄

**Comment by glmnet** [comment]
the docs should mention Color is now Color, not old ESPColor which will work for back compat only.

**Comment by SenexCrenshaw** [comment]
Documents have been added

**Comment by SenexCrenshaw** [comment]
Should I cherry pick https://github.com/esphome/esphome/pull/1395 into this one?

**Comment by glmnet** [comment]
Thanks!

**Pull request merged**